From 2c2f4b90235414fe3a00761d41ada312ba520596 Mon Sep 17 00:00:00 2001
From: Sergey Temerkhanov <stemerkhanov@cavium.com>
Date: Wed, 22 Aug 2018 14:14:57 -0700
Subject: [PATCH 0146/1051] EDAC, thunderx: Maintain platform device symlinks
 for LMC instances

Maintain the device symlink for ThunderX LMC device instances the same
way as it's done for generic EDAC devices

Signed-off-by: Sergey Temerkhanov <s.temerkhanov@gmail.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/edac/thunderx_edac.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/edac/thunderx_edac.c b/drivers/edac/thunderx_edac.c
index 4803c6468bab..491f6d8916f8 100644
--- a/drivers/edac/thunderx_edac.c
+++ b/drivers/edac/thunderx_edac.c
@@ -654,6 +654,8 @@ static inline int pci_dev_to_mc_idx(struct pci_dev *pdev)
 	return ret;
 }
 
+#define LMC_DEVICE_SYMLINK	"device"
+
 static int thunderx_lmc_probe(struct pci_dev *pdev,
 				const struct pci_device_id *id)
 {
@@ -770,6 +772,15 @@ static int thunderx_lmc_probe(struct pci_dev *pdev,
 		goto err_free;
 	}
 
+	ret = sysfs_create_link(&mci->dev.kobj,
+			&mci->pdev->kobj, LMC_DEVICE_SYMLINK);
+
+	if (ret) {
+		dev_err(&pdev->dev, "Cannot create symlink: %d:\n", ret);
+		goto err_del_mc;
+	}
+
+
 	lmc_int = readq(lmc->regs + LMC_INT);
 	writeq(lmc_int, lmc->regs + LMC_INT);
 
@@ -789,6 +800,9 @@ static int thunderx_lmc_probe(struct pci_dev *pdev,
 
 	return 0;
 
+err_del_mc:
+	edac_mc_del_mc(mci->pdev);
+
 err_free:
 	pci_set_drvdata(pdev, NULL);
 	edac_mc_free(mci);
@@ -803,6 +817,7 @@ static void thunderx_lmc_remove(struct pci_dev *pdev)
 
 	writeq(LMC_INT_ENA_ALL, lmc->regs + LMC_INT_ENA_W1C);
 
+	sysfs_remove_link(&mci->dev.kobj, LMC_DEVICE_SYMLINK);
 	edac_mc_del_mc(&pdev->dev);
 	edac_mc_free(mci);
 }
-- 
2.17.1

